proposals/ref-readonly-parameters.md

```diff
@@ -0,0 +1,112 @@
+# `ref readonly` parameters
+
+## Summary
+[summary]: #summary
+
+Allow parameter declaration-site modifier `ref readonly` and change callsite rules as follows:
+
+| Callsite annotation  | `ref` parameter | `ref readonly` parameter | `in` parameter | `out` parameter |
+|----------------------|-----------------|--------------------------|----------------|-----------------|
+| `ref`                | Allowed         | **Allowed**              | **Warning**    | Error           |
+| `in`                 | Error           | **Allowed**              | Allowed        | Error           |
+| `out`                | Error           | **Error**                | Error          | Allowed         |
+| No annotation        | Error           | **Warning**              | Allowed        | Error           |
+
+Note that there is one change to the existing rules: `in` parameter with `ref` callsite annotation produces a warning instead of an error.
+
+| Value kind | `ref` parameter | `ref readonly` parameter | `in` parameter | `out` parameter |
+|------------|-----------------|--------------------------|----------------|-----------------|
+| rvalue     | Error           | **Warning**              | Allowed        | Error           |
```

Comment:
A **Warning** for an rvalue passed for a `ref readonly` parameter sounds really surprising to me. I thought `ref readonly` parameter requires an lvalue as an argument.

Reply:
> This is part of the changes required to allow updating existing signatures in a way that makes more sense, without breaking callers.

I think (and this is just my opinion) this goal renders the feature completely useless for API authors as it doesn't provide for them any real new guarantees. It makes the feature almost useless for consumers as well because warnings can be easily overlooked or quite possibly un-intentionally disabled, or their significance can be misinterpreted. It is quite possible I want a different feature.  

For example, given the current proposal, the following proposed API change doesn't make sense for me:
``` diff
-       public static ReadOnlySpan<T> CreateReadOnlySpan<T>(scoped ref T reference, int length);
+       public static ReadOnlySpan<T> CreateReadOnlySpan<T>(scoped ref readonly T reference, int length);
```

A consumer is now allowed to pass a reference to a clone as the first argument, however, if I was an author of the API, and given its purpose, I would want a situation like that to be disallowed (i.e. impossible). In fact it is disallowed today, because the parameter is a `ref` parameter.

Reply:
There is always contention between the optimal and practical.

For the majority case, it is undesirable to allow passing non location values in and it would be more desirable to error always.

For APIs that are taking `ref` today, the switch to `ref readonly` will be seamless and neither source nor binary breaking.

For APIs that are taking `in` but where they intended to take `ref readonly` and the feature simply didn't exist, then they have no non-breaking path forward as it will result in a source break due to the error.

We then have a few possible paths to handle this, such as implementing a second feature to cover this scenario or slightly relaxing the new language feature to warn instead of error.

The latter seems the lesser of the two evils and avoids needing to duplicate all the same handling the compiler is already doing for `ref readonly` just to raise a different diagnostic kind.